compiler/util/canonicalize_affine.py

```diff
@@ -0,0 +1,121 @@
+from xdsl.ir.affine import (
+    AffineBinaryOpExpr,
+    AffineBinaryOpKind,
+    AffineConstantExpr,
+    AffineDimExpr,
+    AffineExpr,
+    AffineMap,
+)
+
+
+def get_dim(expr: AffineExpr) -> int | None:
+    # get the dimension of this expression. If d0 appears, return 0.
+    # If d1 appears, return 1. This will fail for nested additions in
+    # which multiple dimension expressions appear. This happens quite
+    # often, but can only be solved with a full flattening of the affine map.
+    # this is todo
+    if isinstance(expr, AffineDimExpr):
+        return expr.position
+    if isinstance(expr, AffineBinaryOpExpr):
+        result = get_dim(expr.lhs)
+        if result is not None:
+            return result
+        result = get_dim(expr.rhs)
+        if result is not None:
+            return result
+
+
+def canonicalize_addition(expr: AffineBinaryOpExpr) -> AffineExpr:
+    # always put the constant on rhs
+    assert expr.kind is AffineBinaryOpKind.Add
+    if isinstance(expr.lhs, AffineConstantExpr):
+        # lhs is constant
+        if isinstance(expr.rhs, AffineConstantExpr):
+            # both are constants: fold constants
+            return AffineConstantExpr(expr.lhs.value + expr.rhs.value)
+        else:
+            # move constant to rhs
+            expr = AffineBinaryOpExpr(expr.kind, expr.rhs, expr.lhs)
+    if isinstance(expr.rhs, AffineConstantExpr):
+        # rhs is constant, lhs is not
+        # addition by 0 can be omitted
+        if expr.rhs.value == 0:
+            return expr.lhs
+    # order by minimum dimension first
+    # d0 + d1: good. d1 + d0: bad
+    dim_lhs = get_dim(expr.lhs)
+    dim_rhs = get_dim(expr.rhs)
+    if dim_rhs is not None:
+        if dim_lhs is None or dim_lhs > dim_rhs:
+            return expr.rhs + expr.lhs
+    return expr
```

Comment:
Can you move most of these comments together into a docstring that says everything this function does, instead of it being scattered across this function? I'm giving an example below